internal/discharge: use email to select identity provider

mhilton opened the pull request:
Add an option to the "authentication required" dialogue that allows
a user to choose an identity provider by entering an email address.
This option is only supported by the static and adfs identity provider
types currently.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/59/

mhilton commented on cmd/candid/internal/admincmd/generate-keys.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+var generateKeysDoc = `
+The generate-keys command generates a public/private that can be used in
```
"public/private keypair" perhaps

albertodonato replied:
fixed, thanks

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/60/

albertodonato commented:
can this be landed?

albertodonato commented:
I'll close this as the idea is to have candid generate keys for the config automatically

mhilton committed:
internal/discharge: use email to select identity provider

Add an option to the "authentication required" dialogue that allows
a user to choose an identity provider by entering an email address.
This option is only supported by the static and adfs identity provider
types currently.

cmars commented on idp/static/static.go:
```diff
@@ -76,13 +81,28 @@ func NewIdentityProvider(p Params) idp.IdentityProvider {
+			// if the email address matcher doesn't compile log the error but
+			// carry on. A regular expression that doesn't compile also doesn't
+			// match anything.
+			logger.Errorf("cannot compile match-email-addr regular expression: %s", err)
```
Should this constructor fail and return an error if the configuration is invalid?

mhilton replied:
my thinking is in the comment, but I decided it wasn't worth failing for, instead I decided that nothing can match an invalid regular expression.

cmars commented on idp/openid/openid-connect.go:
```diff
@@ -88,8 +93,22 @@ func NewOpenIDConnectIdentityProvider(params OpenIDConnectParams) idp.IdentityPr
+			// if the email address matcher doesn't compile log the error but
+			// carry on. A regular expression that doesn't compile also doesn't
+			// match anything.
+			logger.Errorf("cannot compile match-email-addr regular expression: %s", err)
```
Same question here

mhilton replied:
same answer too

cmars commented on internal/discharger/login.go:
```diff
@@ -152,17 +156,76 @@ func (h *handler) RedirectLogin(p httprequest.Params, req *redirectLoginRequest)
+// identity provider, or outputs the same email form with an error message.
+func (h *handler) EmailLoginSubmit(p httprequest.Params, req *emailLoginSubmitRequest) error {
+	for _, idp := range h.params.IdentityProviders {
+		type isForEmailAddrer interface {
```
Is this a typo? Should it be `isForEmailAddr`?

mhilton replied:
it's not a typo, but that doesn't mean it's a sensible name.
